PULL_REQUEST_WORKFLOW.rst

```diff
@@ -0,0 +1,260 @@
+slots that are available for the whole organization. More Apache Software Foundation projects started
+to use GitHub Actions and we've started to experience long queues when our jobs waited for free slots.
+
+We approached the problem by:
+
+1) Improving mechanism of canceling duplicate workflow runs more efficiently in case of queue conditions
+   (duplicate workflow runs are generated when someone pushes a fixup quickly - leading to running both
+   out-dated and current run to completion, taking precious slots. This has been implemented by improving
+   `cancel-workflow-run <https://github.com/potiuk/cancel-workflow-runs/>`_ action we are using. In version
+   4.1 it got a new feature of canceling all duplicates even if there is a long queue of builds.
+
+2) Heavily decreasing strain on the Github Actions jobs by introducing selective checks - mechanism
+   to control which parts of the tests are run during the tests. This is implemented by the
+   ``scripts/ci/selective_ci_checks.sh`` script in our repository. This script analyses which part of the
+   code has changed and based on that it sets the right outputs that control which tests are executed in
+   the CI build, and whether we need to build CI images necessary to run those steps. This allowed to
+   heavily decrease the strain especially for the Pull Requests that were not touching code (in which case
+   the builds can complete in < 2 minutes) but also by limiting the number of tests executed in PRs that do
+   not touch the "core" of Airflow, or only touching some - standalone - parts of Airflow such as
+   "Providers", "WWW" or "CLI". This solution is not yet perfect as there are likely some edge cases but
+   it is easy to maintain and we have an escape-hatch - all the tests are always executed in master pushes,
+   so contributors can easily spot if there is a "missed" case and fix it - both by fixing the problem and
+   adding those exceptions to the code. More about it can be found in the
+   `Selective CI checks <#selective-ci-checks>`_ chapter.
+
+3) Even more optimisation came from limiting the scope of tests to only "default" matrix parameters. So far
+   in Airflow we always run all tests for all matrix combinations. The primary matrix components are:
+
+   * Python versions (currently 3.6, 3.7, 3.8)
+   * Backend types (currently MySQL/Postgres)
+   * Backed version (currently MySQL 5.7, MySQL 8, Postgres 9.6, Postgres 13
+
+   We've decided that instead of running all the combinations of parameters for all matrix component we will
+   only run default values (Python 3.6, Mysql 5.7, Postgres 9.6) for all PRs which are not approved yet by
+   the committers. This has the nice effect, that full set of tests (though with limited combinations of
+   the matrix) are still run in the CI for every Pull Request that needs tests at all - allowing the
+   contributors to make sure that their PR is "good enough" to be reviewed.
```

Comment:
```suggestion
   We've decided that instead of running all the combinations of parameters for all matrix component we will
   only run default values (Python 3.6, Mysql 5.7, Postgres 9.6) for all PRs which are not approved yet by
   the committers. This has a nice effect, that full set of tests (though with limited combinations of
   the matrix) are still run in the CI for every Pull Request that needs tests at all - allowing the
   contributors to make sure that their PR is "good enough" to be reviewed.
```